File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefixP(n);
        vector<int> suffixP(n);
        vector<int> r(n);
        prefixP[0] = 1;
        suffixP[n-1] =1;

        for(int i=1;i<n;i++){
            prefixP[i] = nums[i-1]*prefixP[i-1];
        }
        for(int i=n-2;i>=0;i--){
            suffixP[i] = nums[i+1]*suffixP[i+1];
        }
        
        for(int i=0;i<n;i++){
            r[i] = prefixP[i] * suffixP[i];
        }
        return r;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()
        for i, target in enumerate(nums):
            if i>0 and target ==nums[i-1]:
                continue
            l =i+1
            r = len(nums)-1
            while l < r:
                sum1 = target + nums[l] + nums[r]
                if sum1> 0:
                    r -=1
                elif sum1<0:
                    l +=1
                else:
                    ans.append([target,nums[l],nums[r]])
                    l +=1
                    while l<r and nums[l]==nums[l-1]:
                        l+=1
        return ans
        
            